Add support for TalonDoc

wenkokke opened the pull request:
This PR adds support for generating documentation with TalonDoc.

- d68cc92377bc7a4680cfcd8522233c9f1125af03: adds configuration for TalonDoc and Sphinx
- 637304795706e509f5dd04f519111dd526f739df: adds automatically generated documentation

This PR is meant as a starting point. There's a number of considerations:

- Do we want to write Markdown or reStructureText?
  _The PR currently configures Sphinx to use Markdown._
- Do we want to include the generated documentation?
  _The PR currently includes the generated documentation as a starting point._
- Do we want to publish the documentation to GitHub Pages or ReadTheDocs?
  _The PR currently publishes to GitHub Pages. Either requires intervention from someone with administrative privileges._

The PR uses the version 1.0.0 release candidate of TalonDoc, because I anticipate we'll discover some issues while preparing  to merge this PR.

For reference documentation for TalonDoc, see: <https://wen.works/talondoc/>

nriley commented:
@AndreasArvidsson, @phillco and I discussed this briefly on the knausj grooming session today but didn't feel entirely qualified to review it fully.

We had a couple comments:

- Since the .md files contain essentially no useful information, can they be generated somewhere else other than next to the .py/.talon files?
- Is it possible to generate a textual rendering of the documentation instead? If we did want to have documentation "sidecar" files then at least having them be useful to read in a text editor would be great.

I also noted some rendering problems in Safari with https://wen.works/talondoc/ but guessing this may be an issue with one of the downstream tools you are using.

Thanks for doing this!

wenkokke commented:
> Since the generated .md files contain essentially no useful information, can they be generated somewhere else other than next to the .py/.talon files?

Yeah, absolutely. The same folder structure, but under `docs/`?

wenkokke commented:
> Is it possible to generate a textual rendering of the documentation in addition to the HTML version? If we did want to have documentation "sidecar" files then at least having them be useful to read in a text editor would be great.

TalonDoc supports all formats supported by Sphinx, and `talondoc build` is just a wrapper around `sphinx-build -M html`. 

So TalonDoc can generate "HTML (including Windows HTML Help), LaTeX (for printable PDF versions), ePub, Texinfo, manual pages, [and] plain text."

I have no reason to believe the other formats won't work, and you can build them by just calling `sphinx-build` directly.

wenkokke commented:
> I also noted some rendering problems in Safari with https://wen.works/talondoc/ but guessing this may be an issue with one of the downstream tools you are using.

That's on me. I need to debug my CSS, and to be entirely clean should probably move the support for fragmenting to a custom theme. However, this only shows up if you explicitly pass `:width:`.

auscompgeek commented on .github/workflows/ci.yml:
```diff
@@ -12,14 +12,50 @@ permissions:
+
+      - name: Deploy docs
+        if: github.event_name == 'push'
+        uses: JamesIves/github-pages-deploy-action@v4.4.1
```
Given [GitHub Pages supports deploying straight from GitHub Actions](https://github.blog/2022-08-10-github-pages-now-uses-actions-by-default/) rather than pushing build artifacts to a branch, whatever workflow we have here should utilise that.

wenkokke replied:
Would you be happy to push that change to the PR?

auscompgeek replied:
I don't have write access to this PR.

nriley replied:
It seems only maintainers can write to a PR. We may want to fix the other part of this :) — in the meantime, can you attach a patch instead?

pokey replied:
You should be able to file a PR against this PR branch, no?

wenkokke replied:
@auscompgeek I've made those changes. On second thought, nobody push to this branch. The second command adds a large number of files, so I'm going to be forcing this branch to keep it as two separate, clean commits.

wenkokke commented:
I've changed my mind. Comitting all the generated files is an incredible amount of noise for very little benefit, so I've removed them. Instead, the CI generates these files, and publishes them to pages, and I will instead work on making the autogen output better. In the meantime, we can work on writing high quality docs for knausj_talon without the burden of all these generated files.

auscompgeek commented on CONTRIBUTING.md:
```diff
@@ -1,13 +1,63 @@
+
+For community-generated documentation on Talon, please visit https://talon.wiki/
+
+## Alternate installation method: Zip file
```
I think this section was meant to be for non-contributors

wenkokke replied:
Copied back to README

wenkokke commented:
@knausj85 I think I need you to merge this, because it requires configuring Pages for the repository.

C-Loftus commented:
I was curious with the status of this PR. Is this paused for a technical/logistical reason? Seems like it could be quite useful and it seems to be all done. Just wasn't sure if I was overlooking something. Was planning to add some docs to one of my own repos and wasn't sure if this would be the preferred way.

wenkokke commented:
While this PR was being prepared, the community repository adopted a novel Talon feature that isn't yet supported by TalonDoc. That's the current blocker.

nriley commented:
@wenkokke on the community backlog session, we were reviewing old PRs and wondering if this is something you plan to pick up again in future or if this is something one of us should consider taking over? There continues to be a substantial need for better community documentation. Thanks!

wenkokke commented:
> @wenkokke on the community backlog session, we were reviewing old PRs and wondering if this is something you plan to pick up again in future or if this is something one of us should consider taking over? There continues to be a substantial need for better community documentation. Thanks!

The current version works, but not incredibly well. This PR was foiled by the release of Talon lists, which aren't currently supported. This means that the information that was previously read from CSV files (and therefore showed up in TalonDoc) is now absent, because TalonDoc doesn't know how to read Talon list files.

nriley commented:
> @knausj85 I think I need you to merge this, because it requires configuring Pages for the repository.

@knausj85 From the community backlog session — we agree that it makes sense to pick this up and try to get it merged even if it doesn't currently support Talon lists. Would you be able to make this configuration change for the community repository? The rest of us can't work on getting it merged after that.

We also discussed that there is not currently a standard for documentation strings in community. It would make sense to add something to CONTRIBUTING.md to help explain our standards once we get some…

wenkokke commented:
> > @knausj85 I think I need you to merge this, because it requires configuring Pages for the repository.
> 
> 
> 
> @knausj85 From the community backlog session — we agree that it makes sense to pick this up and try to get it merged even if it doesn't currently support Talon lists. Would you be able to make this configuration change for the community repository? The rest of us can't work on getting it merged after that.
> 
> 
> 
> We also discussed that there is not currently a standard for documentation strings in community. It would make sense to add something to CONTRIBUTING.md to help explain our standards once we get some…

That request predates the move to talonhub, so I think any current talonhub admin or owner could do this.

nriley commented:
Thanks! I'm out of time for Talon stuff this weekend anyway.
